smart_open/tests/test_s3_version.py

```diff
@@ -0,0 +1,104 @@
+    boto3.resource('s3').BucketVersioning(BUCKET_NAME).enable()
+
+
+def put_to_bucket(contents, num_attempts=12, sleep_time=5):
+    # fake (or not) connection, bucket and key
+    logger.debug('%r', locals())
+
+    #
+    # In real life, it can take a few seconds for the bucket to become ready.
+    # If we try to write to the key while the bucket while it isn't ready, we
+    # will get a ClientError: NoSuchBucket.
+    #
+    for attempt in range(num_attempts):
+        try:
+            boto3.resource('s3').Object(BUCKET_NAME, KEY_NAME).put(Body=contents)
+            return
+        except botocore.exceptions.ClientError as err:
+            logger.error('caught %r, retrying', err)
+            time.sleep(sleep_time)
+
+    assert False, 'failed to create bucket %s after %d attempts' % (BUCKET_NAME, num_attempts)
+
+
+@maybe_mock_s3
+class TestVersionId(unittest.TestCase):
+    def setUp(self):
+        self.test_ver0 = u"String version 1.0".encode('utf8')
+        self.test_ver1 = u"String version 2.0".encode('utf8')
+        # write into key
+        with smart_open.s3.BufferedOutputBase(BUCKET_NAME, WRITE_KEY_NAME) as fout:
+            fout.write(self.test_ver0)
+        with smart_open.s3.BufferedOutputBase(BUCKET_NAME, WRITE_KEY_NAME) as fout:
+            fout.write(self.test_ver1)
+
+    def test_good_id(self):
+        """Does version_id into s3 work correctly?"""
+        versions = boto3.resource('s3').Bucket(BUCKET_NAME).object_versions.filter(Prefix=WRITE_KEY_NAME)
+        check_version = list(versions)[0].get()['VersionId']
+        with smart_open.s3.SeekableBufferedInputBase(BUCKET_NAME, WRITE_KEY_NAME, check_version) as fin:
+            expected = fin.read()
+        self.assertEqual(expected, self.test_ver0)
+
+    def test_bad_id(self):
+        """Does version_id exception into s3 work correctly?"""
+        versions = boto3.resource('s3').Bucket(BUCKET_NAME).object_versions.filter(Prefix=WRITE_KEY_NAME)
+        check_version = list(versions)[0].get()['VersionId']
+
+        with self.assertRaises(IOError):
+            smart_open.s3.open(BUCKET_NAME, WRITE_KEY_NAME, 'rb', version_id='bad-version-does-not-exist')
+
+    def test_bad_mode(self):
+        """Check the write mode selection when transferring the version"""
+        versions = boto3.resource('s3').Bucket(BUCKET_NAME).object_versions.filter(Prefix=WRITE_KEY_NAME)
+        check_version = list(versions)[0].get()['VersionId']
+
+        with self.assertRaises(ValueError):
+            smart_open.s3.open(BUCKET_NAME, WRITE_KEY_NAME, 'wb', version_id=check_version)
+
+    def test_top_level(self):
+        "Checks top level function open"
+        versions = boto3.resource('s3').Bucket(BUCKET_NAME).object_versions.filter(Prefix=WRITE_KEY_NAME)
+        check_version = list(versions)[0].get()['VersionId']
+        with open("s3://"+BUCKET_NAME+"/"+WRITE_KEY_NAME, 'rb', check_version) as fin:
+            expected = fin.read()
```

Comment:
This test is missing assertions. You need to check the following:

1. Passing in no version at all gives the newest version of the file
2. Passing in the newest version explicitly gives the same as above
3. Passing in the oldest version gives you the oldest version

Get ready for a surprise ;)